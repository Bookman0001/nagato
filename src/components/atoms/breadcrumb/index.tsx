import styled from 'styled-components'
import Link from 'next/link'

import { BreadcrumbContents } from 'src/types'

interface Props {
  contents: BreadcrumbContents
}

export default function Breadcrumb({ contents }: Props) {
  const contentsLength = contents.length
  return (
    <Container>
      {contents.map((item, index) => {
        return (
          <Link href={item.link} key={index}>
            <Item>
              {item.label}
              {contentsLength !== index && <span>{'>'}</span>}
            </Item>
          </Link>
        )
      })}
    </Container>
  )
}

const Container = styled.div`
  display: flex;
`

const Item = styled.div`
  cursor: pointer;
  span {
    margin: 0 5px;
  }
`
