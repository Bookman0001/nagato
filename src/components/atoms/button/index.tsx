import React, { ButtonHTMLAttributes, MouseEvent } from 'react'
import styled from 'styled-components'

import { BORDER_RADIUS, COLOR, FONT_SIZE } from 'src/theme/constants'

interface Props extends ButtonHTMLAttributes<HTMLButtonElement> {
  disabled?: boolean
  children?: React.ReactNode
  onClick?: () => void
}

export default function Button({ disabled, children, onClick }: Props) {
  const handleClick = (e: MouseEvent) => {
    if (onClick) {
      e.stopPropagation()
      e.preventDefault()
      onClick()
    }
  }
  return (
    <ButtonItem onClick={handleClick} disabled={disabled} role={'button'}>
      {children}
    </ButtonItem>
  )
}

const ButtonItem = styled.button<{ width?: string }>`
  min-height: 40px;
  height: 100%;
  background-color: ${COLOR.WHITE};
  border: 2px solid ${COLOR.BLACK};
  border-radius: ${BORDER_RADIUS.DEFAULT};
  padding: 0 1rem;
  display: flex;
  justify-content: center;
  align-items: center;
  font-size: ${FONT_SIZE.M};
  cursor: pointer;
  :hover {
    opacity: 0.7;
  }
  :disabled {
    opacity: 0.7;
  }
  @media (prefers-color-scheme: dark) {
    background-color: ${COLOR.BLACK};
    border: 2px solid ${COLOR.WHITE};
    color: ${COLOR.WHITE};
  }
`
