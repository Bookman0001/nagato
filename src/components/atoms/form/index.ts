export { Input } from './input'
export { Label } from './label'
export { Select } from './select'
export { TextArea } from './textArea'
