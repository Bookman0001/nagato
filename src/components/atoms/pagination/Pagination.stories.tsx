import { useState, VFC } from 'react'
import { Meta } from '@storybook/react/types-6-0'

import { Pagination } from './'

export default {
  title: 'Pagination',
  component: Pagination,
} as Meta

const Template: VFC = () => {
  const [count, setCount] = useState<number>(1)
  return (
    <>
      総件数50件で10件ずつの時
      <Pagination
        totalCount={50}
        limit={10}
        onClick={(index) => setCount(index)}
        currentIndex={count}
      />
    </>
  )
}

export const Default = Template.bind({})
