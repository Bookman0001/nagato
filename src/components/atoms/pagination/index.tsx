import styled, { css } from 'styled-components'

import { COLOR } from '../../../theme/constants'

interface Props {
  currentIndex: number
  totalCount: number
  limit: number
  onClick: (clickedIndex: number) => void
}

export default function Pagination({
  currentIndex,
  totalCount,
  limit,
  onClick,
}: Props) {
  const paginationCount = Math.ceil(totalCount / limit)

  const handleClick = (clickedIndex: number) => {
    if (clickedIndex > limit) {
      return
    }
    onClick(clickedIndex)
  }

  return (
    <Container>
      {[...Array(paginationCount)].map((_, index) => {
        return (
          <ButtonWrapper
            key={index}
            hasSolidBorder={currentIndex === index + 1}
          >
            <ButtonItem onClick={() => handleClick(index + 1)} key={index}>
              {index + 1}
            </ButtonItem>
          </ButtonWrapper>
        )
      })}
    </Container>
  )
}

const Container = styled.div`
  display: flex;
  justify-content: center;
  align-items: center;
`

const ButtonWrapper = styled.div<{ hasSolidBorder: boolean }>`
  button {
    border: 1px solid ${COLOR.BLACK};
  }
  ${(props) =>
    props.hasSolidBorder &&
    css`
      button {
        border: 2px solid ${COLOR.BLACK};
      }
    `}
`

const ButtonItem = styled.button`
  display: block;
  background-color: ${COLOR.WHITE};
  border-radius: 10%;
  width: 30px;
  height: 30px;
  margin: 0 1rem;
  border: none;
  font-size: 0.75rem;
  cursor: pointer;
  text-align: center;
`
