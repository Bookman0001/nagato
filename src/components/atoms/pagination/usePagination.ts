import { useCallback } from 'react'

export function usePagination({
  currentIndex,
  totalCount,
  limit,
}: {
  currentIndex: number
  totalCount: number
  limit: number
}) {
  const distanceCountFromEdge = 3

  const totalPaginationCount = Math.ceil(totalCount / limit)
  const onceShowingCountLength = distanceCountFromEdge * 2 + 1
  const defaultShowingCountLength =
    onceShowingCountLength < totalPaginationCount
      ? onceShowingCountLength
      : totalPaginationCount

  /* eslint-disable max-statements */
  const getDisplayPaginations = useCallback(() => {
    let end = currentIndex + distanceCountFromEdge
    let start =
      currentIndex - distanceCountFromEdge > 0
        ? currentIndex - distanceCountFromEdge
        : 1
    if (currentIndex <= distanceCountFromEdge + 1) {
      end = defaultShowingCountLength
    }
    if (currentIndex + distanceCountFromEdge >= totalPaginationCount) {
      end = totalPaginationCount
      if (
        totalPaginationCount - currentIndex < distanceCountFromEdge &&
        totalPaginationCount - currentIndex >= 0
      ) {
        start = totalPaginationCount - (defaultShowingCountLength - 1)
      }
    }

    return Array(end - start + 1)
      .fill(null)
      .map((_, i) => i + start)
  }, [currentIndex, defaultShowingCountLength, totalPaginationCount])
  /* eslint-enable max-statements */

  return {
    getDisplayPaginations,
  }
}
