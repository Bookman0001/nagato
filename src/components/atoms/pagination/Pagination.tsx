import { useMemo } from 'react'
import styled, { css } from 'styled-components'

import { usePagination } from './usePagination'

import { borderRadius, color, fontSize } from 'src/theme/constants'

interface Props {
  currentIndex: number
  totalCount: number
  limit: number
  onClick: (clickedIndex: number) => void
}

export function Pagination({
  currentIndex,
  totalCount,
  limit,
  onClick,
}: Props) {
  const { getDisplayPaginations } = usePagination({
    currentIndex,
    totalCount,
    limit,
  })

  const displayPaginations = useMemo(
    () => getDisplayPaginations(),
    [getDisplayPaginations]
  )

  if (!totalCount) return null

  const handleClick = (clickedIndex: number) => {
    onClick(clickedIndex)
  }

  return (
    <Container>
      {displayPaginations.map((num, index) => {
        return (
          <ButtonWrapper key={index} hasSelected={currentIndex === num}>
            <ButtonItem
              onClick={() => handleClick(num)}
              key={index}
              role={'button'}
            >
              {num}
            </ButtonItem>
          </ButtonWrapper>
        )
      })}
    </Container>
  )
}

const Container = styled.div`
  display: flex;
  justify-content: center;
  align-items: center;
`

const ButtonWrapper = styled.div<{ hasSelected: boolean }>`
  ${(props) =>
    props.hasSelected &&
    css`
      button {
        color: ${color.BLACK};
        background-color: ${color.WHITE};
        border: 3px solid ${color.WHITE};
      }
    `}
`

const ButtonItem = styled.button`
  display: flex;
  justify-content: center;
  align-items: center;
  background-color: ${color.BLACK};
  border: 2px solid ${color.WHITE};
  border-radius: ${borderRadius.DEFAULT};
  width: 45px;
  height: 45px;
  margin: 0 1rem;
  font-size: ${fontSize.S};
  cursor: pointer;
  text-align: center;
`
