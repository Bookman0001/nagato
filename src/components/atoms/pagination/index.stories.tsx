import React from 'react'
import { Meta } from '@storybook/react/types-6-0'
import { action } from '@storybook/addon-actions'

import Pagination from './'

export default {
  title: 'Pagination',
  component: Pagination,
} as Meta

const Template: React.VFC<{}> = () => (
  <>
    総件数50件で10件ずつの時
    <Pagination
      totalCount={50}
      initialIndex={1}
      limit={10}
      onClick={action('clicked')}
    />
  </>
)

export const Default = Template.bind({})
