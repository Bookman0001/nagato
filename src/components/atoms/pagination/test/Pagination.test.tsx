import { render, screen, fireEvent } from '@testing-library/react'

import { Pagination } from 'src/components/atoms/pagination'

describe('Pagination', () => {
  beforeEach(() => {
    jest.clearAllMocks()
  })

  it('should be rendered', () => {
    const handleClick = jest.fn()
    render(
      <Pagination
        currentIndex={1}
        totalCount={10}
        limit={2}
        onClick={handleClick}
      />
    )
    expect(screen.getAllByRole('button')).toBeDefined()
    fireEvent.click(screen.getAllByRole('button')[0])
    expect(handleClick).toHaveBeenCalledTimes(1)
  })

  it('should not to be rendered', () => {
    const handleClick = jest.fn()
    const { container } = render(
      <Pagination
        currentIndex={1}
        totalCount={0}
        limit={2}
        onClick={handleClick}
      />
    )
    expect(container.firstChild).toBe(null)
  })
})
