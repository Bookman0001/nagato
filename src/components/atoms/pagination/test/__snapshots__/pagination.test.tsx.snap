// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pagination 1`] = `
<styled.div>
  <styled.div
    hasSolidBorder={false}
    key="0"
  >
    <styled.button
      key="0"
      onClick={[Function]}
    >
      1
    </styled.button>
  </styled.div>
  <styled.div
    hasSolidBorder={false}
    key="1"
  >
    <styled.button
      key="1"
      onClick={[Function]}
    >
      2
    </styled.button>
  </styled.div>
</styled.div>
`;
